### PR TITLE
Fix: Viewport Normal operation & Shaking when Resize

### LINKED ISSUE #2783: `scroll_with_delta` doesn't work
**Describe the bug**

`scroll_with_delta` doesn't seem to work.

**To Reproduce**

1. Create an app with `eframe`.
2. Copy the `scroll_with_delta` example from documentation.
3. Click `Scroll down` button.
4. See that nothing happens.

```rust
use eframe::egui::{self, Vec2};

fn main() -> Result<(), eframe::Error> {
    eframe::run_native(
        "My egui App",
        eframe::NativeOptions::default(),
        Box::new(|_cc| Box::new(MyApp {})),
    )
}

struct MyApp {}

impl eframe::App for MyApp {
    fn update(&mut self, ctx: &egui::Context, _frame: &mut eframe::Frame) {
        egui::CentralPanel::default().show(ctx, |ui| {
            let mut scroll_delta = Vec2::ZERO;
            if ui.button("Scroll down").clicked() {
                scroll_delta.y -= 64.0; // move content up
            }
            egui::ScrollArea::vertical().show(ui, |ui| {
                ui.scroll_with_delta(scroll_delta);
                for i in 0..1000 {
                    ui.label(format!("Item {}", i));
                }
            });
        });
    }
}
```

**Expected behavior**

Content scrolls down.

**Screenshots**

[video](https://user-images.githubusercontent.com/12128106/222928092-28c4f85e-29e7-4b11-9992-f4811a4c41ad.webm)

**Desktop (please complete the following information):**

 - OS: GNU/Linux

### LINKED ISSUE #3811: Window not destroyed on exit on windows
**Describe the bug**
On windows, when clicking the x in the titlebar the window isn't destroyed if the application doesn't exit. This seems similar to #2892. The window closes as expected on version 0.23.0 but fails on 0.25.0 (and also on the current master branch)

**To Reproduce**
In a loop open a window using `eframe::run_native` with the default options on windows. Close the window. It will now persist until the next call to run_native (at that point the old window is destroyed and a new one created).

```rust
fn main() -> Result<()> {
 loop {
    let options = eframe::NativeOptions::default();
    let _ = eframe::run_simple_native("test", options, |_cc, _frame| {});
    std::thread::sleep(std::time::Duration::from_secs(5));
  }
}
```

**Desktop (please complete the following information):**
 - OS: Windows 11

### LINKED ISSUE #3902: 5% cpu usage After sending ViewportCommand::Visible (false)
The following text was generated by automatic translation:
**Describe the bug**
If focus on the egui window,Continuously produces the following events after sending ViewportCommand::Visible (false).
```
winit event: NewEvents(Poll)
winit event: AboutToWait
```
The program has 5% usage in the background, On a Ryzen 5900x with 12 cores and 24 threads.
This does not happen if the window is not in focus or after it is minimized

**Expected behavior**
Same cpu usage as window minimization.

**Desktop (please complete the following information):**
 - OS: Windows 10  LTSC 19044.3693
 - Version:  master

### LINKED ISSUE #3953: App window lags when dragged around the screen on Windows 10
Hi, I'm using **Windows 10** (I have **8GB** of **RAM** and **Intel i5**) and the _latest version of Rust, egui and eframe_ (I added them 5 minutes ago) and when I'm experiencing an issue that on I couldn't see on the [video](https://www.youtube.com/watch?v=NtUkr_z7l84).

Basically, when I `cargo run`, the window appears after 20 seconds of loading (I don't know if the wait is normal whenever I make a change on the code), when I drag the window it **lags** like there is a delay in frames.
You can see what happens in the video, I opened notepad and as you can see when I drag the window, it is smooth. Why does this happen with the rust one??

https://github.com/emilk/egui/assets/84017765/3c8fe878-f066-4a87-8bb5-5433f4e91589

### LINKED ISSUE #3972: When the program is minimized, the viewport stops working.
When a new Immediate Viewport is opened and the program is minimized,
the new Immediate Viewport remains displayed on the screen but becomes unresponsive.
When the program is restored from the minimized state, the Immediate Viewport starts functioning again.

This issue occurs after a recent update.
In previous versions, the new Immediate Viewport was not displayed on the screen when the program was minimized.

OS: Window 10
egui version: master branches

### LINKED ISSUE #4022: Immediate Viewport speed issue
Viewport speed issue

If you run `Immediate Viewport` while the program is at 60 frames (per second),
it will be at 50 frames.

If you have `two monitors` and the Viewport is on a different screen from the program,
the frame rate will be 30.

Is this a normal reaction?

egui version : ALL
OS : Windows 10

### LINKED ISSUE #4070: The program screen flickers (shakes) when resizing.
The program screen flickers (shakes) when resizing.

OS: Windows 10
egui Version : ALL, master

### LINKED ISSUE #4135: Infinite negative dimensions on window
The windows created by `eframe` on my system seem to not be able to detect their dimensions, which I noticed due to `ViewportBuilder::with_min_inner_size` not doing anything. (To potentially save time for people reading this, I suspect this is due to my somewhat bleeding-edge desktop environment, which I included details about below in the relevant section).

**To Reproduce**

1. Put the following contents into the respective files:

```toml
# Cargo.toml
[package]
edition = "2021"
name = "scratch"
version = "0.1.0"

# This happens both on the latest published version and when depending directly on the latest master branch
[dependencies]
#eframe = "0.26.2"
eframe = { git = "https://github.com/emilk/egui", branch = "master" }
```

```rust
# src/main.rs
use eframe::{egui::ViewportBuilder, NativeOptions};

fn main() {
    eframe::run_simple_native(
        "is this dark matter?",
        NativeOptions {
            viewport: ViewportBuilder::default().with_min_inner_size([1000.0, 1000.0]),
            ..Default::default()
        },
        |ctx, _| {
            dbg!(ctx.used_size(), ctx.used_rect());
        },
    )
    .expect("wait that wasn't supposed to be what went wrong...");
}
```

2. `cargo run` or `cargo run --release`

3. Check that the window is rendered properly

4. Check the print output and see the following:

```rust
[src/main.rs:11:13] ctx.used_size() = [-inf -inf]
[src/main.rs:11:13] ctx.used_rect() = [[inf inf] - [-inf -inf]]
[src/main.rs:11:13] ctx.used_size() = [-inf -inf]
[src/main.rs:11:13] ctx.used_rect() = [[inf inf] - [-inf -inf]]
[src/main.rs:11:13] ctx.used_size() = [-inf -inf]
[src/main.rs:11:13] ctx.used_rect() = [[inf inf] - [-inf -inf]]
[src/main.rs:11:13] ctx.used_size() = [-inf -inf]
[src/main.rs:11:13] ctx.used_rect() = [[inf inf] - [-inf -inf]]
...
```

**Expected behavior**

The dimensions reported by `Context::used_rect` and `Context::used_size` would be accurate (presumably positive, given that the window does seem to render fine) and the minimum size specified would be respected.

**Screenshots**

image

**Desktop:**

I suspect this might be where the issues lies; I'm using Plasma 6 from the Arch Linux testing repos, which only became available this week.

 - OS: Linux (details from `uname` shown below)

```
→ uname -a
Linux bunchacrunch 6.7.8-zen1-1-zen #1 ZEN SMP PREEMPT_DYNAMIC Sun, 03 Mar 2024 00:30:23 +0000 x86_64 GNU/Linux
```

 - Wayland (see attached output of `wayland-info` for details)

[wayland_info.txt](https://github.com/emilk/egui/files/14484298/wayland_info.txt)

I'm unfortunately not sure what the best way to determine which rendering output is being used by eframe. I'm fairly certain that it's not using X11 because I generally can identify windows using XWayland by using `wmctrl` because it doesn't ever list info about native Wayland windows, and it similarly doesn't have any output related to the window opened by `eframe`. I'm not sure if it's using the `wayland` backend or `glow` (or if those are entirely different things rather than mutually exclusive options; I'm pretty inexperienced in programming GUIs).

Assuming this isn't reproducible easily on other desktops, I'm happy to add any additional info (either from my system or updating the repro code to print something else) that would be helpful to try to narrow cause what's causing this! If there's a different destkop environment that would be helpful for me to check, it shouldn't be too much of an issue; reverting back to Plasma 5 to test this would unfortunately be a bit of a pain given that I plan to keep using 6, but if there's no other way to sanity check things, I might be able to find an old laptop or something to boot up and try it out on that.

### LINKED ISSUE #4178: There is on very rare occasions a case where a panic occurred
There is on very rare occasions a case where a panic occurred at line 2,132 of context.rs when a Viewport exit.
( On screens with different pixels_per_point. )

### LINKED ISSUE #4179: There is on very rare occasions a case where a panic occurred at line 152 of layers.rs
There is on very rare occasions a case where a panic occurred at line 152 of layers.rs when a Viewport exit.
( At this time, self.0.len() == 0 )

( On screens with different pixels_per_point. )

thread 'main' panicked at C:\prog\egui\crates\egui\src\layers.rs:152:15:
index out of bounds: the len is 0 but the index is 0

### LINKED ISSUE #4216: ViewportInfo.parent doesn't always store any information.
`ViewportInfo.parent` doesn't always store any information.

and

It would be nice to have `ViewportInfo.this`.

### LINKED ISSUE #4222: For each Viewport, to store `transparent` and `decoration` information.
For each Viewport, it is a good idea to store `transparent` and `decoration` information.

### PR DESCRIPTION
**Pull request by rustbasic**
This will continue to be updated :

Issues :
* Closes #2783 
* Closes #3811 
* Closes #3902 
* Closes #3953
* Closes #3972 
* Closes #4022
* Closes #4070 
* Related #4091
* Closes #4135 
* Related #4173 
* Closes #4178 
* Closes #4179
* Closes #4216 
* Closes #4222 

Pull Requests :
* Related #4040 
* Related #4043 
* Related #4054 
* Related #4071 
* Closes #4088
* Closes #4154
* Related #4214 
* Closes #4217
* Closes #4242 
* Closes #4328 
* Closes #4332 
* Closes #4348 
* Closes #4360
* Closes #4362 
* Closes #4374
* Closes #4388  
* Closes #4412 
* Related #4487 
* Related #4488 
* Related #4508
* Closes #4509

This will continue to be updated :

1. It may not be merged, so use it for reference only.
2. It is always in a usable state, maintaining a state with few bugs.
3. Anyone is free to maintain or improve a portion of this, and Pull Request it.

Based on `examples/test_viewports` and `examples/multiple_viewports`, all fixed to work properly.
`wgpu` only tested `examples/test_viewports`, so additional testing on your part is needed.

### CODE REVIEW

**Comment by rustbasic** [comment]
After (v0.25.0), the part that always seemed to call 'request_repaint()' was changed to repaint only when an event occurs.

=> change again : epi_intergration.rs 293 line
```
// TODO : Do not recall until the next repaint.
request_repaint_after_for(std::time::Duration::from_millis(8), viewport_id)
```

File: crates/eframe/src/native/epi_integration.rs [hunk 1]
```diff
@@ -278,13 +278,20 @@ impl EpiIntegration {
...
+                // ROOT Viewport ( with Immediate Viewport )
                 crate::profile_scope!("App::update");
                 app.update(egui_ctx, &mut self.frame);
             }
+
+            let viewport_id = egui_ctx.viewport_id();
+            // TODO : Do not recall until the next repaint.
+            // 1000 millis / 60 fps = 16.67 millis
+            self.egui_ctx
+                .request_repaint_after_for(std::time::Duration::from_millis(8), viewport_id);
```
Comment (emilk) [inline crates/eframe/src/native/epi_integration.rs:294]: this will just cause a hot repaint loop

```suggestion
```

File: crates/egui/src/widgets/text_edit/builder.rs [hunk 2]
```diff
@@ -413,15 +413,15 @@ impl<'t> TextEdit<'t> {
...
+                        ui.visuals().widgets.unhovered.bg_stroke, // TODO(emilk): we want to show something here, or a text-edit field doesn't "pop".
                     )
                 }
             } else {
                 let visuals = &ui.style().visuals.widgets.inactive;
                 epaint::RectShape::stroke(
                     frame_rect,
                     visuals.rounding,
-                    visuals.bg_stroke, // TODO(emilk): we want to show something here, or a text-edit field doesn't "pop".
+                    ui.visuals().widgets.unhovered.bg_stroke, // TODO(emilk): we want to show something here, or a text-edit field doesn't "pop".
```
Comment (emilk) [inline crates/egui/src/widgets/text_edit/builder.rs:424]: please remove unrelated commits

File: crates/egui-winit/src/lib.rs [hunk 3]
```diff
@@ -939,21 +939,49 @@ pub fn update_viewport_info(
...
+    // Deprecated
     if cfg!(target_os = "windows") {
         // It's tempting to do this, but it leads to a deadlock on Mac when running
         // `cargo run -p custom_window_frame`.
         // See https://github.com/emilk/egui/issues/3494
-        viewport_info.maximized = Some(window.is_maximized());
         viewport_info.minimized = Some(window.is_minimized().unwrap_or(false));
+        viewport_info.maximized = Some(window.is_maximized());
     }
+    */
```
Comment (emilk) [inline crates/egui-winit/src/lib.rs:958]: remove dead code

File: crates/egui-winit/src/lib.rs [hunk 4]
```diff
@@ -939,21 +939,49 @@ pub fn update_viewport_info(
...
+
+    update_info.minimized = match cfg!(target_os = "windows") {
+        true => Some(window.is_minimized().unwrap_or(false)),
+        false => window.is_minimized(),
```
Comment (emilk) [inline crates/egui-winit/src/lib.rs:962]: As the comment above says, this leads to deadlock on Mac

File: crates/eframe/src/native/winit_integration.rs [hunk 5]
```diff
@@ -106,7 +106,9 @@ pub enum EventResult {
 
     RepaintAt(WindowId, Instant),
 
-    Exit,
+    ViewportExit(WindowId),
+
+    Exit(WindowId),
```
Comment (emilk) [inline crates/eframe/src/native/winit_integration.rs:111]: Needs docstrings

File: crates/eframe/src/native/run.rs [hunk 6]
```diff
@@ -227,6 +240,7 @@ fn run_and_return(
             })
             .ok();
     }
+    */
```
Comment (emilk) [inline crates/eframe/src/native/run.rs:243]: Remove dead code

**Review by emilk — CHANGES_REQUESTED** [review]
Very exciting if this indeed closes this many issues!

File: crates/egui-winit/src/lib.rs (hunk 4 above)
Comment (rustbasic) [reply crates/egui-winit/src/lib.rs]: > As the comment above says, this leads to deadlock on Mac

Okay, I was curious about the results.

**Comment by emilk** [comment]
Converted to draft until unrelated commits are removed

**Comment by rustbasic** [comment]
> Converted to draft until unrelated commits are removed

Yes, since it is a mixture of several things, it seems that division will be possible only after several commits are completed.

**Comment by emilk** [comment]
this PR is too big
